debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlStreamingChangeEventSource.java

```diff
@@ -212,7 +214,25 @@ protected void onEvent(MySqlOffsetContext offsetContext, Event event) {
             return;
         }
 
-        ts = clock.currentTimeInMillis() - eventTs;
+        // Prefer higher resolution replication timestamps from MySQL 8 GTID events, if possible
+        if (isGtidModeEnabled) {
+            if (event.getHeader().getEventType() == EventType.GTID) {
+                GtidEventData gtidEvent = unwrapData(event);
+                final long gtidEventTs = gtidEvent.getOriginalCommitTimestamp();
+                if (gtidEventTs == 0) {
+                    // < MySQL 8.0.1, fallback to second resolution timestamp from the event header
+                    eventTimestamp = Instant.ofEpochMilli(eventTs);
+                } else {
+                    // >= MySQL 8.0.1, prefer the higher resolution replication timestamp
+                    eventTimestamp = Instant.EPOCH.plus(gtidEventTs, ChronoUnit.MICROS);
+                }
+            }
+        } else {
+            // Fallback to second resolution event timestamps
+            eventTimestamp = Instant.ofEpochMilli(eventTs);
+        }
```

Comment:
May I suggest refactoring this block into a separate method and having only two statements, the one that determines the ">= MySQL 8.0.1" behavior returning the higher precision timestamp or the fallback that handles all other cases?

Reply:
Thanks for the review Chris, on it.